### DISCUSSION TITLE
Why isn't Text contained in the JSON string?

### DISCUSSION THREAD

**Discussion by shynur** [capnproto/capnproto#2192] [Q&A] [answered]
Here's the capnp file:

```
struct MessageHeader {
  channel   @0 :Text;
  timestamp @1 :UInt64;
}
```

Here's the test function (C++):

```cpp
void test_capnproto() {
    auto header = capnp::MallocMessageBuilder{}.initRoot<MessageHeader>();
    header.setChannel("123"),  // <-- where is "123"?
    header.setTimestamp(996);

    auto json = capnp::JsonCodec{}.encode(header);
    std::cerr << "Cap'n: " << json.cStr() <<'\n';
}
```

It printed only `Cap'n: {"timestamp":"5"}`.

Can any kind person tell me why?  Thank you.

**Comment by kentonv** [comment, marked as answer]
This line:

```
    auto header = capnp::MallocMessageBuilder{}.initRoot<MessageHeader>();
```

creates a temporary `MallocMessageBuilder`, which is destroyed at the end of the statement. `initRoot` returns a _pointer_ (or rather, a poniter-like object) to the root of the message, but after this line of code, the pointer is left dangling. The next few lines are modifying and then stringifying an object that no longer exists, which is undefined behavior. This could segfault, but in your case it apparently just corrupted memory instead, and produced a garbage result.

To fix this, you need to allocate `MallocMessageBuilder` as a local variable that stays alive until you're done with it:

```
    capnp::MallocMessageBuilder message;
    auto header = message.initRoot<MessageHeader>();
```
